web/src/message_view.ts

```diff
@@ -974,6 +974,30 @@ export function maybe_add_local_messages(opts: {
             // below), but the server doesn't support that query.
             id_info.final_select_id = id_info.target_id;
         }
+
+        // If we have cached data for the same narrow, we can try to find the
+        // message we want to select locally.
+        if (superset_data.filter.equals(filter) && !load_local_messages(msg_data, superset_data)) {
+            return;
+        }
+
+        // If we locally have the message user wants to select, then
+        // we can render locally.
+        if (
+            id_info.target_id &&
+            msg_data.get(id_info.target_id) &&
+            !filter.can_mark_messages_read()
+        ) {
+            id_info.local_select_id = id_info.target_id;
+        } else if (
+            id_info.target_id === undefined &&
+            // Our goal here is to select the first unread id. We can only do
+            // so if we have the complete data for the narrow.
+            msg_data.fetch_status.has_found_newest() &&
+            msg_data.fetch_status.has_found_oldest()
+        ) {
+            id_info.local_select_id = msg_data.first_unread_message_id();
+        }
```

Comment:
Hmm, can you explain why this logic is being added to the `cannot_compute` block? I'm not sure what assumptions are being made here that justify being able to navigate the user correctly. (I could imagine maybe that we should instead be not returning `cannot_compute` if we have the view cached and can somehow use that to pin the right message ID to select).

Reply:
So, here if we have the `MessageListData` with the same filter as we are trying to render, we can try to render the
message list locally.

We have a `target_id` we want to select - Likely user searched for something -> clicked on a messsage and pressed back key. Then, we can select the message.

We don't have a `target_id` to select but have the complete message history for the filter. In that case we can just render and select the first unread.

Reply:
OK. I may want to add some comments here.

Reply:
Done